File: data-structures/single-linked-list/nodeList.js
const Node = require("./node.js");
class NodeList {
  constructor() {
    this.head = null;
  }
  printNodes() {
    var cursor = new Node();
    cursor = this.head;
    while (cursor.next != null) {
      console.log(cursor);
      cursor = cursor.next;
    }
    console.log(cursor);
  }
  insertAtEnd(newNode) {
    if (newNode === null) {
      throw new Error("Can't add null node");
    }
    if (this.head === null) {
      this.head = newNode;
      return this.head;
    }
    var cursor = new Node();
    cursor = this.head;
    while (cursor.next != null) {
      cursor = cursor.next;
    }
    cursor.next = newNode;
  }
  insertInFront(newNode) {
    if (newNode === null) {
      throw new Error("Can't add null node");
    }
    if (this.head === null) {
      this.head = newNode;
      return this.head;
    }
    newNode.next = this.head;
    this.head = newNode;
  }
  insertAtIndex(index, newNode) {
    if (newNode === null) {
      throw new Error("Can't add null node");
    }
    if (this.head === null && index === 0) {
      this.head = newNode;
    } else if (index === 0) {
      insertAtFront(newNode);
    }
    var cursor = new Node();
    cursor = this.head;
    while ((index-- > 0) && (cursor.next !== null)) {
      cursor = cursor.next;
    }
    // if the cursor can't go anymore and we are still not at indexing, this index is out of range
    if (cursor.next === null && index === 0) {
      throw new Error("Can't add node at invalid index");
    }
    newNode.next = cursor.next;
    cursor.next = newNode;
  }
  getTail() {
    var cursor = new Node();
    cursor = this.head;
    while (cursor.next != null) {
      cursor = cursor.next;
    }
    return cursor;
  }
}
module.exports = NodeList;

File: data-structures/single-linked-list/nodeList.cpp
#include "node.cpp"
#include <iostream>

using namespace std;

class NodeList {
	struct Node* head;
	public:
		NodeList();
		void printNodes();
		struct Node* insertInFront(struct Node*);
		struct Node* insertAtEnd(struct Node*);
		struct Node* insertAtIndex(int, struct Node*);
		struct Node* getTail();
};

NodeList::NodeList() {
	head = NULL;
}

void NodeList::printNodes() {
	struct Node* ptr;

	for (ptr = head; ptr != NULL; ptr = ptr->next)	
		cout << ptr->data << "\n";
}

struct Node* NodeList::insertInFront(struct Node* newNode) {
	if (newNode == NULL)
		return NULL;
 
	if (head == NULL) {
		head = newNode;
		newNode->next = NULL;
		return head;
	}
 	
 	newNode->next = head;
 	head = newNode;
 	return head;
}

struct Node* NodeList::insertAtEnd(struct Node* newNode) {
	struct Node* ptr;

	if (newNode == NULL) 
		return NULL;

	if (head == NULL) {
		head = newNode;
		newNode->next = NULL;
		return head;
	}

	ptr = head;
	while (ptr->next != NULL) {
		ptr = ptr->next;
	}

	ptr->next = newNode;
	return newNode;
}

struct Node* NodeList::insertAtIndex(int index, struct Node* newNode) {
	struct Node* ptr;

	if (newNode == NULL || index < 0)
		return NULL;

	if (head == NULL && index == 0) {
		head = newNode;
		return head;
	} else if (index == 0) {
		return insertInFront(newNode);
	} else {
		ptr = head;
		while (--index > 0) 
			ptr = ptr->next;
		newNode->next = ptr->next;
		ptr->next = newNode;
		return newNode;
	}
}

struct Node* NodeList::getTail() {
	struct Node* ptr;
	for (ptr = head; ptr->next != NULL; ptr = ptr->next);
	return ptr;
}